### PR TITLE
EN-7961/Storage-cache-for-smart-contracts

### PR DESCRIPTION
**Pull request by sasurobert**
Created storage and cache for compiled smart contract code

### CODE REVIEW

File: process/smartContract/hooks/blockChainHook.go [hunk 1]
```diff
@@ -480,6 +494,42 @@ func (bh *BlockChainHookImpl) SetCurrentHeader(hdr data.HeaderHandler) {
...
+// SaveCompiledCode saves to cache and storage the compiled code
+func (bh *BlockChainHookImpl) SaveCompiledCode(codeHash []byte, code []byte) {
+	bh.compiledScPool.Put(codeHash, code, len(code))
+	err := bh.compiledScStorage.Put(codeHash, code)
+	if err != nil {
+		log.Debug("SaveCompiledCode", "error", err)
+	}
+}
+
+// GetCompiledCode returns the compiled code if it finds in the cache or storage
```
Comment (SebastianMarian) [inline process/smartContract/hooks/blockChainHook.go:506]: if it is found

File: process/smartContract/hooks/blockChainHook.go [hunk 2]
```diff
@@ -480,6 +494,42 @@ func (bh *BlockChainHookImpl) SetCurrentHeader(hdr data.HeaderHandler) {
...
+			return found, compiledCode
+		}
+	}
+
+	compiledCode, err := bh.compiledScStorage.Get(codeHash)
+	if err != nil {
+		return false, nil
+	}
+
+	return found, compiledCode
```
Comment (SebastianMarian) [inline process/smartContract/hooks/blockChainHook.go:521]: return true, compiledCode as "found" could be false here

File: cmd/node/config/config.toml [hunk 3]
```diff
@@ -195,6 +195,18 @@
         MaxBatchSize = 20000
         MaxOpenFiles = 10
 
+[SmartContractsStorage]
+    [SmartContractsStorage.Cache]
+        Name = "SmartContractsStorage"
+        Capacity = 1000
+        Type = "LRU"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/config/config.toml:202]: done

File: dataRetriever/factory/dataPoolFactory.go [hunk 4]
```diff
@@ -90,6 +90,13 @@ func NewDataPoolFromConfig(args ArgsDataPool) (dataRetriever.PoolsHolder, error)
 		return nil, err
 	}
 
+	cacherCfg = factory.GetCacherFromConfig(mainConfig.SmartContractDataPool)
+	smartContracts, err := storageUnit.NewCache(cacherCfg)
+	if err != nil {
+		log.Info("error creating trieNodes")
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/factory/dataPoolFactory.go:96]: done

File: process/smartContract/hooks/blockChainHook.go (hunk 1 above)
Comment (sasurobert) [reply process/smartContract/hooks/blockChainHook.go:506]: done

File: process/smartContract/hooks/blockChainHook.go (hunk 2 above)
Comment (sasurobert) [reply process/smartContract/hooks/blockChainHook.go:521]: done

**Review by SebastianMarian — DISMISSED** [review]

File: process/smartContract/hooks/blockChainHook.go [hunk 5]
```diff
@@ -64,14 +69,23 @@ func NewBlockChainHookImpl(
...
+		blockChain:        args.BlockChain,
+		shardCoordinator:  args.ShardCoordinator,
+		marshalizer:       args.Marshalizer,
+		uint64Converter:   args.Uint64Converter,
+		builtInFunctions:  args.BuiltInFunctions,
+		compiledScPool:    args.DataPool.SmartContracts(),
+		compiledScStorage: args.StorageService.GetStorer(dataRetriever.SmartContractUnit),
+	}
+
+	if check.IfNil(blockChainHookImpl.compiledScStorage) {
```
Comment (bogdan-rosianu) [inline process/smartContract/hooks/blockChainHook.go:84]: you could move these checks before assigning the values to the `blockChainHookImpl`

File: process/smartContract/hooks/blockChainHook.go [hunk 6]
```diff
@@ -480,6 +494,42 @@ func (bh *BlockChainHookImpl) SetCurrentHeader(hdr data.HeaderHandler) {
 	bh.mutCurrentHdr.Unlock()
 }
 
+// SaveCompiledCode saves to cache and storage the compiled code
```
Comment (bogdan-rosianu) [inline process/smartContract/hooks/blockChainHook.go:497]: maybe rephrase to `// SaveCompiledCode saves the compiled code to cache and storage`

File: process/smartContract/hooks/blockChainHook.go [hunk 7]
```diff
@@ -480,6 +494,42 @@ func (bh *BlockChainHookImpl) SetCurrentHeader(hdr data.HeaderHandler) {
...
+func (bh *BlockChainHookImpl) SaveCompiledCode(codeHash []byte, code []byte) {
+	bh.compiledScPool.Put(codeHash, code, len(code))
+	err := bh.compiledScStorage.Put(codeHash, code)
+	if err != nil {
+		log.Debug("SaveCompiledCode", "error", err)
```
Comment (bogdan-rosianu) [inline process/smartContract/hooks/blockChainHook.go:502]: you can also log the `codeHash` for debug reasons

File: process/smartContract/hooks/blockChainHook.go [hunk 8]
```diff
@@ -480,6 +494,42 @@ func (bh *BlockChainHookImpl) SetCurrentHeader(hdr data.HeaderHandler) {
...
+	}
+}
+
+// GetCompiledCode returns the compiled code if it is found in the cache or storage
+func (bh *BlockChainHookImpl) GetCompiledCode(codeHash []byte) (bool, []byte) {
+	val, found := bh.compiledScPool.Get(codeHash)
+	if found {
+		compiledCode, ok := val.([]byte)
+		if ok {
+			return found, compiledCode
```
Comment (bogdan-rosianu) [inline process/smartContract/hooks/blockChainHook.go:512]: maybe `return true, compiledCode`? it's obvious that `found` is true but it would be more clear if you would use `true` directly

File: process/smartContract/hooks/blockChainHook.go [hunk 9]
```diff
@@ -480,6 +494,42 @@ func (bh *BlockChainHookImpl) SetCurrentHeader(hdr data.HeaderHandler) {
...
+
+	compiledCode, err := bh.compiledScStorage.Get(codeHash)
+	if err != nil {
+		return false, nil
+	}
+
+	return true, compiledCode
+}
+
+// DeleteCompiledCode deletes from storage and cache the compiled code
```
Comment (bogdan-rosianu) [inline process/smartContract/hooks/blockChainHook.go:524]: deletes the compiled code from storage and cache*

File: process/smartContract/hooks/blockChainHook.go [hunk 10]
```diff
@@ -480,6 +494,42 @@ func (bh *BlockChainHookImpl) SetCurrentHeader(hdr data.HeaderHandler) {
...
+func (bh *BlockChainHookImpl) DeleteCompiledCode(codeHash []byte) {
+	bh.compiledScPool.Remove(codeHash)
+	err := bh.compiledScStorage.Remove(codeHash)
+	if err != nil {
+		log.Debug("DeleteCompiledCode", "error", err)
```
Comment (bogdan-rosianu) [inline process/smartContract/hooks/blockChainHook.go:529]: also log the code hash

File: dataRetriever/interface.go [hunk 11]
```diff
@@ -83,6 +83,8 @@ const (
 	MiniblockHashByTxHashUnit UnitType = 14
 	// ReceiptsUnit is the receipts storage unit identifier
 	ReceiptsUnit UnitType = 15
+	// SmartContractUnit is the smart contracts storage unit identifier
+	SmartContractUnit UnitType = 16
```
Comment (bogdan-rosianu) [inline dataRetriever/interface.go:87]: please adapt the comment on line 82 as to match the function's name

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: process/smartContract/hooks/blockChainHook.go (hunk 5 above)
Comment (sasurobert) [reply process/smartContract/hooks/blockChainHook.go:84]: it is the same. It is less code like this.

File: process/smartContract/hooks/blockChainHook.go (hunk 6 above)
Comment (sasurobert) [reply process/smartContract/hooks/blockChainHook.go:497]: done

File: process/smartContract/hooks/blockChainHook.go (hunk 7 above)
Comment (sasurobert) [reply process/smartContract/hooks/blockChainHook.go:502]: this will happen only if we have a problem with storage - codehash does not help. but added.

File: process/smartContract/hooks/blockChainHook.go (hunk 8 above)
Comment (sasurobert) [reply process/smartContract/hooks/blockChainHook.go:512]: done

File: process/smartContract/hooks/blockChainHook.go (hunk 9 above)
Comment (sasurobert) [reply process/smartContract/hooks/blockChainHook.go:524]: done

File: process/smartContract/hooks/blockChainHook.go (hunk 10 above)
Comment (sasurobert) [reply process/smartContract/hooks/blockChainHook.go:529]: done

File: dataRetriever/interface.go (hunk 11 above)
Comment (sasurobert) [reply dataRetriever/interface.go:87]: done

File: process/smartContract/hooks/blockChainHook.go (hunk 7 above)
Comment (bogdan-rosianu) [reply process/smartContract/hooks/blockChainHook.go]: yes, but at least we can know that for a given hash the save failed

**Review by bogdan-rosianu — APPROVED** [review]

File: dataRetriever/interface.go [hunk 12]
```diff
@@ -79,7 +79,7 @@ const (
 	MiniblocksMetadataUnit UnitType = 12
 	// EpochByHashUnit is the epoch by hash storage unit identifier
 	EpochByHashUnit UnitType = 13
-	// MiniblocksHashByTxHashUnit is the miniblocks hash by tx hash storage unit identifier
+	// MiniblockHashByTxHashUnit is the miniblocks hash by tx hash storage unit identifier
```
Comment (SebastianMarian) [inline dataRetriever/interface.go:82]: "is the miniblock hash" instead "is the miniblocks hash"

**Review by SebastianMarian — APPROVED** [review]
